### PR TITLE
Jasp refactoring (Part 2)

### PR DESCRIPTION
**Pull request by PietropaoloFrisoni**
## Description

I worked with Claude to remove a lot of boilerplate, dead code, code repetitions and perform code improvements (mostly based on `pylance`, `pylint` and `ruff`, as usual) across the Jasp infrastructure . We also catched and fixed some bugs during the process. Review should be hopefully easy as this is basically a massive internal refactoring with no breaking changes.

Removes duplicated and dead code across Qrisp's `jasp` (JAX-tracing) subsystem: the same control-flow, equation-copying, and caching boilerplate had been independently copy-pasted across the Catalyst, classical-boolean-simulation, profiling, and post-processing interpreter backends. Everything beyond the dedup itself is static-analysis cleanup (pylint/pyright/ruff) surfaced while touching these files, plus the handful of real bugs that surfaced along the way.

## Related Issues

Closes #
Related to #

## Type of Change

- [ ] Feature (new functionality)
- [ ] Change Request (modification of existing functionality)
- [x] Bug Fix
- [x] Refactoring (no behavior change) — primary focus of this PR
- [ ] Performance improvement
- [ ] Documentation
- [ ] CI / Build

## Breaking Change?

- [x] Yes
- [ ] No

`stimulate()`'s signature narrowed from `stimulate(func: Callable | None = None)` to `stimulate(func: Callable)`. In practice this is not expected to affect anyone: the `None` default was already dead (no handling existed for it, so `@stimulate()` with empty parens would have failed downstream anyway), and no call site anywhere in the codebase used the zero-arg form. Migration path: use `@stimulate` (no parentheses), which is how it was already always used.

## What was changed?

**Centralization (the main point of this PR):**
- Unified the "replay `while`/`scan` under an active trace" logic (previously duplicated across the Catalyst, classical-boolean-simulation, profiling, and post-processing interpreters) into `evaluate_while_loop_under_trace`/`evaluate_scan_under_trace` in `control_flow_interpretation.py`; same for `cond` into `evaluate_cond_under_trace` where the two consumers shared the same strategy.
- Unified `flatten_signature`/`unflatten_signature` (Catalyst vs. classical-boolean-simulation interpreters) and the `copy_jaxpr_eqn`/`_copy_eqn` equation-copying helper (previously reimplemented 5 times across `environment_flattening.py`, `composite_gate_interpreter.py`, `inv_transform.py`, `injection_transform.py`, and `conjugation_environment.py`).
- New shared helpers replacing other scattered duplicates: `insert_call_outvalues` (call-equation output wrapping, 4 sites), `build_metric_profiler`/`get_cached_jaspr` (profiling-metric boilerplate), `rebuild_jaxpr` (new `jasp_expression/jaxpr_utils.py`, 3 sites), `_wrap_branch_for_tracing`/`_get_last_cond_eqn`/`_finalize_branch_eqn` (`q_cond`/`q_switch`), `tracing_scope` (exception-safe tracing context manager, `qache`/`quantum_kernel`), `_bump_gate_count` (`BufferedQuantumState`), `_decompose_and_exec` (composite-gate decomposition), `_register_vars`/`_slice_field` (`VarTracker`), `get_op_counts` (gate-count breakdown).
- Removed dead code: `Jaspr.inline()` (zero callers anywhere, now delegates to `embedd(inline=True)`), a redundant simulator-validation branch in `simulate_jaspr` (identical error already raised one line later by `BufferedQuantumState.__init__`).

**Bug fixes found while centralizing:**
- `evaluate_scan_under_trace`: a `while`/`scan` with exactly one carried value crashed (carry/pytree-structure mismatch) across all four consumer backends — the duplicated copies all shared this bug.
- `extract_post_processing`: a `KeyError` when a jaxpr-level constant (e.g. a `scan`'s `xs` array) was referenced, from consts never being bound into the evaluation context.
- `AbstractQubitArray._getitem`: an off-by-one slice-bound error (dead code, no reachable caller, but now shares correct logic with the live `DynamicQubitArray.__getitem__` path).

**Static-analysis cleanup:** type hints, pylint findings, and pyright errors resolved across `primitives/`, `jaspification.py`, `boolean_simulation.py`, `BufferedQuantumState`, and the control-flow interpreter files — no intended behavior change. Two latent type-hint bugs (`Jaspr.debug_info` wrongly `Optional`, `Jaspr.count_ops`/`.depth`/`.num_qubits` declaring `meas_behavior: str` when `str | Callable` is actually accepted) were corrected as part of this.

**New tests:** coverage added for previously-untested lines across `test_basic_primitives.py`, `test_boolean_simulation.py`, `test_buffered_quantum_state.py` (new), `test_catalyst_interface.py`, `test_count_ops.py`, `test_jaspify.py`, `test_post_processing_extraction.py`, `test_stim_simulation.py` (new) — including regression tests for the num-carry-one scan bug across all four affected backends.

## How was it tested?

No issue-linked Test-IDs apply here (this PR isn't tied to a specific bug-tracker issue), so this table describes verification method rather than Test-IDs:

| Verification | Status |
|---------|--------|
| New/updated pytest tests (8 files, listed above) | ✅ |
| Static analysis: pyright, pylint, ruff (format + lint) on every touched file | ✅ |
| Fresh-process import checks after every change to eagerly-loaded modules | ✅ |
| Manual functional verification of the centralization batch not yet covered by pytest (see "further tests to add" in reviewer notes) | ✅ (manual only) |

## Screenshots / Output (if applicable)

N/A — internal refactor, no user-visible output changes.

## Checklist
- [x] My code follows the project's coding standards
- [x] I have performed a self-review
- [ ] I have added/updated tests (referencing issue Test-IDs) — partial: see reviewer notes, the most recent centralization batch has manual-only verification
- [ ] All tests pass locally and in CI — please confirm in CI; two test failures found post-push (exact-match error-message assertions) have been fixed
- [ ] I have updated the documentation — not needed; `stimulate`'s doc page is Sphinx `autofunction`-generated from the signature
- [ ] I have added a changelog entry — this repo has no CHANGELOG file to update
- [x] Breaking changes are documented with migration path

## Reviewer Notes

- The `stimulate()` signature change (see "Breaking Change?" above) is the only public-API-visible change in this PR; everything else is internal interpreter/tracing-tool plumbing.
- Worth a close look: `evaluate_scan_under_trace`'s num-carry-one fix, since it changes behavior for a previously-crashing case across four backends at once (Catalyst, classical simulation, profiling, post-processing) — regression tests were added for all four.
- Not yet covered by automated tests (manual verification only): the `q_cond`/`q_switch` state-stripped branch path, `tracing_scope`'s exception path, `get_cached_jaspr`'s cache-hit path. Happy to add these before merge if preferred.
- A follow-up PR could pick up ~11 more duplication-removal opportunities identified during the same audit (not included here to keep this PR reviewable).

### CODE REVIEW

File: src/qrisp/jasp/interpreter_tools/interpreters/profiling_interpreter.py [hunk 1]
```diff
@@ -32,21 +32,33 @@
 """
 
 
+import types
 from abc import ABC, abstractmethod
 from collections import OrderedDict
-from typing import Any, Callable, Dict, Sequence, Tuple
+from typing import TYPE_CHECKING, Any, Callable, Dict, Sequence, Tuple
```
Comment (renezander90) [inline src/qrisp/jasp/interpreter_tools/interpreters/profiling_interpreter.py]:
```suggestion
from collections.abc import Callable, Sequence
from typing import TYPE_CHECKING, Any, Dict, Tuple
```

File: src/qrisp/jasp/interpreter_tools/interpreters/cl_func_interpreter.py [hunk 2]
```diff
@@ -15,44 +15,43 @@
...
+#    - Gates creating superposition (H, RX, RY, etc.) raise an error.
+#
+# The transformation works by:
+# 1. Replacing quantum primitives with classical bit manipulation operations
+# 2. Converting control flow (while, cond) to use the classical reinterpretation
+# 3. Producing a standard ClosedJaxpr that can be JIT-compiled by JAX
 
 from collections import OrderedDict
+from collections.abc import Sequence
 from typing import Any, Callable
```
Comment (renezander90) [inline src/qrisp/jasp/interpreter_tools/interpreters/cl_func_interpreter.py]:
```suggestion
from collections.abc import Callable, Sequence
from typing import Any
```

File: src/qrisp/jasp/interpreter_tools/interpreters/count_ops_metric.py [hunk 3]
```diff
@@ -15,21 +15,19 @@
 ********************************************************************************
 """
 
-import types
 from typing import Callable, Dict, List, Tuple
```
Comment (renezander90) [inline src/qrisp/jasp/interpreter_tools/interpreters/count_ops_metric.py]:
```suggestion
from collections.abc import Callable
from typing import Dict, List, Tuple
```

File: src/qrisp/jasp/interpreter_tools/interpreters/depth_metric.py [hunk 4]
```diff
@@ -15,7 +15,6 @@
 ********************************************************************************
 """
 
-import types
 from typing import Any, Callable, Iterator, List, Tuple
```
Comment (renezander90) [inline src/qrisp/jasp/interpreter_tools/interpreters/depth_metric.py]:
```suggestion
from collections.abc import Callable, Iterator
from typing import Any, List, Tuple
```

File: src/qrisp/jasp/interpreter_tools/interpreters/num_qubits_metric.py [hunk 5]
```diff
@@ -15,19 +15,16 @@
 ********************************************************************************
 """
 
-import types
 from typing import Callable, Tuple
```
Comment (renezander90) [inline src/qrisp/jasp/interpreter_tools/interpreters/num_qubits_metric.py]:
```suggestion
from collections.abc import Callable
from typing import Tuple
```

File: src/qrisp/jasp/interpreter_tools/interpreters/control_flow_interpretation.py [hunk 6]
```diff
@@ -217,3 +235,254 @@ def evaluate_scan(scan_eq, context_dic, eqn_evaluator=exec_eqn):
...
+    Evaluates a JAX cond equation by reinterpreting each branch Jaxpr with the given
+    eqn_evaluator and replaying it via ``jax.lax.switch``, preserving the branch
+    structure under an active trace.
+
+    Unlike :func:`evaluate_cond_eqn`, which picks and evaluates a single branch eagerly
+    using a concrete Python index (for use outside of tracing mode), this function is
+    meant to be called *while* the interpreter itself is being traced -- the resulting
+    switch primitive is a real traced JAX operation.
+
+    Args:
```
Comment (renezander90) [inline src/qrisp/jasp/interpreter_tools/interpreters/control_flow_interpretation.py]: Deviates from the NumPy style used throughout the Qrisp documentation.

File: src/qrisp/jasp/interpreter_tools/interpreters/control_flow_interpretation.py [hunk 7]
```diff
@@ -217,3 +235,254 @@ def evaluate_scan(scan_eq, context_dic, eqn_evaluator=exec_eqn):
...
+    with the given eqn_evaluator and replaying them via ``jax.lax.while_loop``,
+    preserving the loop structure under an active trace.
+
+    Unlike :func:`evaluate_while_loop`, which unrolls the loop eagerly using concrete
+    Python truthiness (for use outside of tracing mode), this function is meant to be
+    called *while* the interpreter itself is being traced (e.g. to compile a profiling
+    metric, or to build a post-processing Jaxpr) -- the resulting while_loop primitive
+    is a real traced JAX operation.
+
+    Args:
```
Comment (renezander90) [inline src/qrisp/jasp/interpreter_tools/interpreters/control_flow_interpretation.py]: Deviates from the NumPy style used throughout the Qrisp documentation.

File: src/qrisp/jasp/interpreter_tools/interpreters/control_flow_interpretation.py [hunk 8]
```diff
@@ -217,3 +235,254 @@ def evaluate_scan(scan_eq, context_dic, eqn_evaluator=exec_eqn):
...
+    Evaluates a JAX scan equation by reinterpreting the body Jaxpr with the given
+    eqn_evaluator and replaying it via ``jax.lax.scan``, preserving the loop
+    structure under an active trace.
+
+    Unlike :func:`evaluate_scan`, which unrolls the scan eagerly with a plain Python
+    for-loop (for use outside of tracing mode), this function is meant to be called
+    *while* the interpreter itself is being traced -- the resulting scan primitive is
+    a real traced JAX operation.
+
+    Args:
```
Comment (renezander90) [inline src/qrisp/jasp/interpreter_tools/interpreters/control_flow_interpretation.py]: Deviates from the NumPy style used throughout the Qrisp documentation.

File: src/qrisp/jasp/interpreter_tools/abstract_interpreter.py [hunk 9]
```diff
@@ -15,7 +15,7 @@
 ********************************************************************************
 """
 
-from typing import Callable, Sequence
+from typing import Any, Callable
```
Comment (renezander90) [inline src/qrisp/jasp/interpreter_tools/abstract_interpreter.py]:
```suggestion
from collections.abc import Callable
from typing import Any
```

File: src/qrisp/jasp/interpreter_tools/interpreters/control_flow_interpretation.py (hunk 6 above)
Comment (PietropaoloFrisoni) [reply src/qrisp/jasp/interpreter_tools/interpreters/control_flow_interpretation.py]: Thanks, I should tell my friend Claude to use the same convention as well (this are not user-facing functions, but it is worth using the same convention everywhere)

**Review by renezander90 — APPROVED** [review]
LGTM 🎊

**Comment by positr0nium** [comment]
Originally, the control flow interpretation file was meant to host static interpreters (i.e. turning Jax control flow into Python control flow). I yield that the name is maybe not super telling here. I think that it might still make sense to have a separation between this case and the things that you added. Could you move the traced control flow interpretation into a more suitable location, @PietropaoloFrisoni ?

**Comment by PietropaoloFrisoni** [comment]
I moved the traced-control-flow interpreters into a new file, so that `control_flow_interpretation.py` only has the static Python-control-flow versions. Let me know if you'd prefer a different name or split :)

**Review by positr0nium — APPROVED** [review]
Thanks :)
